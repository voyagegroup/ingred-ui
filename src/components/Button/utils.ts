import { Theme } from "../../themes";
import { ButtonColor, ButtonColorStyle } from "./Button";

export const getContainerColorStyles = (
  theme: Theme,
): { [P in ButtonColor]: ButtonColorStyle } => ({
  primary: {
    normal: {
      background: theme.palette.primary.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowPrimary"],
      border: `1px solid ${theme.palette.primary.dark}`,
    },
    hover: {
      background: theme.palette.primary.dark,
      border: "none",
    },
    active: {
      background: theme.palette.primary.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  basicLight: {
    normal: {
      background: theme.palette.background.default,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.divider}`,
    },
    hover: {
      background: theme.palette.gray.highlight,
      border: `1px solid ${theme.palette.divider}`,
    },
    active: {
      background: theme.palette.gray.highlight,
      boxShadow: theme.shadow["3dShadowActive"],
      border: `1px solid ${theme.palette.divider}`,
    },
  },
  basicDark: {
    normal: {
      background: theme.palette.basicDark.main,
      color: theme.palette.black,
      boxShadow: theme.shadow["3dShadowBasic"],
      border: `1px solid ${theme.palette.basicDark.deepDark}`,
    },
    hover: {
      background: theme.palette.basicDark.dark,
      border: "none",
    },
    active: {
      background: theme.palette.basicDark.dark,
      boxShadow: theme.shadow["3dShadowBasicActive"],
      border: "none",
    },
  },
  danger: {
    normal: {
      background: theme.palette.danger.main,
      color: theme.palette.text.white,
      boxShadow: theme.shadow["3dShadowDanger"],
      border: `1px solid ${theme.palette.danger.dark}`,
    },
    hover: {
      background: theme.palette.danger.dark,
      border: "none",
    },
    active: {
      background: theme.palette.danger.dark,
      boxShadow: theme.shadow["3dShadowActive"],
      border: "none",
    },
  },
  clear: {
    normal: {
      background: "none",
      color: theme.palette.black,
      boxShadow: "none",
      border: "none",
    },
    hover: {
      background: theme.palette.gray.light,
      border: "none",
    },
    active: {
      background: theme.palette.gray.main,
      boxShadow: "none",
      border: "none",
    },
  },
  primaryPale: {
    normal: {
      background: theme.palette.primaryPale.main,
      color: theme.palette.primaryPale.dark,
      boxShadow: theme.shadow["3dShadowPrimaryPale"],
      border: `1px solid ${theme.palette.primaryPale.light}`,
    },
    hover: {
      background: theme.palette.primaryPale.highlight,
      border: "none",
    },
    active: {
      background: theme.palette.primaryPale.highlight,
      boxShadow: theme.shadow["3dShadowPrimaryPaleActive"],
      border: "none",
    },
  },
});
