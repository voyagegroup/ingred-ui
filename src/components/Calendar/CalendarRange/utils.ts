import dayjs, { Dayjs } from "dayjs";

const isStart = (startDate: Dayjs | null, month: Dayjs, day: number) =>
  startDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isEnd = (endDate: Dayjs | null, month: Dayjs, day: number) =>
  endDate?.format("YYYY-MM-DD") ===
  dayjs(new Date(month.year(), month.month(), day)).format("YYYY-MM-DD");

const isBetween = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) =>
  (startDate &&
    endDate &&
    dayjs(new Date(month.year(), month.month(), day)).isAfter(
      startDate.format("YYYY-MM-DD"),
      "day",
    ) &&
    dayjs(new Date(month.year(), month.month(), day)).isBefore(
      endDate.format("YYYY-MM-DD"),
      "day",
    )) ??
  false;

export const getDayState = (
  startDate: Dayjs | null,
  endDate: Dayjs | null,
  month: Dayjs,
  day: number,
) => {
  if (isStart(startDate, month, day)) {
    return "start";
  }
  if (isEnd(endDate, month, day)) {
    return "end";
  }
  if (isBetween(startDate, endDate, month, day)) {
    return "between";
  }
  return "none";
};
