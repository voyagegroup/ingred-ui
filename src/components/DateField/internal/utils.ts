export const getInputWidth = (input: string) =>
  [...input.split("")].reduce((acc, char) => {
    const code = char.charCodeAt(0);
    return acc + (code >= 0x3000 && code <= 0xff60 ? 2 : 1);
  }, 0);
