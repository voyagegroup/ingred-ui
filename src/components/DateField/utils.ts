import { useMemo } from "react";

type Sections = {
  start: number;
  end: number;
  value: string;
  editable: boolean;
};

/**
 * 何らかのフォーマットで入ってくる日付を開始位置と終了位置と値を持つセクションに分割する
 * useDateField で format された日付操作を汎用的に行うために必要なプロパティを返す
 * 例) 2023-01-02 -> [
 *   { start: 0, end: 3, value: "2023", editable: true },
 *   { start: 4, end: 4, value: "-", editable: false },
 *   { start: 5, end: 6, value: "01", editable: true },
 *   { start: 7, end: 7, value: "-", editable: false },
 *   { start: 8, end: 9, value: "02", editable: true }
 * ]
 *
 * @param formattedDate 何らかのフォーマットで入ってくる日付
 * @returns 開始位置と終了位置と値を持つセクション
 */
export const getSections = (formattedDate?: string | null) => {
  if (!formattedDate) {
    return [];
  }

  const sections: Sections[] = [];
  let start = 0;
  let isPrevCharDigit = !isNaN(Number(formattedDate[0]));

  for (let index = 1; index <= formattedDate.length; index++) {
    const currentChar = formattedDate[index];
    const isCurrentCharDigit =
      !isNaN(Number(currentChar)) && currentChar !== " ";

    if (
      isCurrentCharDigit !== isPrevCharDigit ||
      currentChar === " " ||
      index === formattedDate.length
    ) {
      sections.push({
        start,
        end: index - 1,
        value: formattedDate.slice(start, index),
        editable: isPrevCharDigit,
      });
      start = index;
      isPrevCharDigit = isCurrentCharDigit;
    }
  }

  return sections;
};

/**
 * 開始位置と終了位置と値を持つセクションをフォーマットされた日付に変換する
 */
export const formatString = (sectionsWithCharactor: Sections[]) =>
  sectionsWithCharactor.map((section) => section.value).join("");

type ReactRef<T> =
  | React.RefCallback<T>
  | React.MutableRefObject<T>
  | React.ForwardedRef<T>
  | string
  | null
  | undefined;

// from: https://github.com/voyagegroup/ingred-ui/blob/master/src/hooks/useMergeRefs.ts
export function useMergeRefs<T>(...refs: ReactRef<T>[]): React.Ref<T> {
  return useMemo(() => {
    if (refs.every((ref) => ref === null)) {
      return null;
    }
    return (refValue: T) => {
      for (const ref of refs) {
        if (typeof ref === "function") {
          ref(refValue);
        } else if (ref && typeof ref !== "string") {
          ref.current = refValue;
        }
      }
    };
  }, [refs]);
}
