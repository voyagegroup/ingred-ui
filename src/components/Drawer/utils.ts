/**
 * 相対値をpx値に変換するヘルパー関数
 * @param value - 変換する値（数値またはvw/vh/%/px等の単位付き文字列）
 * @param direction - Drawerの方向（変換時の基準値決定に使用）
 * @returns px値（数値）
 */
export const convertToPixels = (value: string | number, direction: "left" | "right" | "bottom"): number => {
  if (typeof value === "number") {
    return value; // 数値はpx扱い
  }
  
  // 文字列の場合は単位チェック
  const numericValue = parseFloat(value);
  
  if (value.includes("vw")) {
    return (numericValue / 100) * window.innerWidth;
  }
  if (value.includes("vh")) {
    return (numericValue / 100) * window.innerHeight;
  }
  if (value.includes("%")) {
    // %の場合は親要素基準だが、ここではviewport基準として扱う
    const isHorizontal = direction === "left" || direction === "right";
    return (numericValue / 100) * (isHorizontal ? window.innerWidth : window.innerHeight);
  }
  if (value.includes("px")) {
    return numericValue;
  }
  
  // 単位なしの文字列は警告を出す
  if (!isNaN(numericValue)) {
    console.warn(`Drawer: 文字列サイズには単位を指定してください。"${value}" → "${value}px" を推奨`);
    return numericValue; // 一応数値として扱う
  }
  
  // 無効な値の場合はエラー警告とフォールバック
  console.error(`Drawer: 無効なサイズ値です: "${value}". デフォルト値を使用します。`);
  return 400; // フォールバック
}; 