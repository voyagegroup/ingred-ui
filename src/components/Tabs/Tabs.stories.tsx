import * as React from "react";
import Tabs from ".";

export default {
  title: "Components/Navigation/Tabs",
  component: Tabs,
};

export const Example = () => {
  const [value, setValue] = React.useState("日別");
  const options = {
    data: [
      { text: "日別", value: "日別" },
      { text: "月別", value: "月別" },
    ],
  };

  return <Tabs {...options} value={value} onChange={setValue} />;
};

export const WithBadge = () => {
  const [value, setValue] = React.useState("全て");
  const options = {
    data: [
      { text: "全て", count: 5, value: "全て" },
      { text: "ユニット", count: 5, value: "ユニット" },
      { text: "サイズ", count: 5, value: "サイズ" },
    ],
  };

  return (
    <Tabs {...options} value={value} withBadge={true} onChange={setValue} />
  );
};
