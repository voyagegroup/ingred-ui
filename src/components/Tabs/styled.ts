import styled from "styled-components";
import Typography from "../Typography";

type ButtonProps = {
  selected: boolean;
};

export const ChildContainer = styled.div`
  position: relative;
`;

export const ChildrenContainer = styled.div``;

export const Button = styled.button<ButtonProps>`
  flex-irection: column;
  flex-shrink: 0;
  line-height: 1.25;
  max-width: 360;
  min-height: 48;
  min-width: 90;
  overflow: hidden;
  margin: 0;
  padding: 12px 16px;
  position: relative;
  text-align: center;
  white-space: normal;
  border: none;
  outline: none;
  background-color: transparent;
  cursor: ${({ selected }) => (selected ? "" : "pointer")};
  &:hover {
    background-color: ${({ selected }) =>
      !selected ? ({ theme }) => theme.palette.gray.highlight : ""};
    transition: 0.5s;
  }
`;

export const Text = styled(Typography)<ButtonProps>`
  padding-right: 2px;
  font-weight: 600;
  color: ${({ selected }) =>
    selected ? ({ theme }) => theme.palette.primary.main : "gray"};
`;

export const Indicator = styled.div`
  position: absolute;
  top: 0;
  height: 100%;
  transition: all 0.3s ease-in-out;
  border-bottom: ${({ theme }) => `solid ${theme.palette.primary.main} 2px`};
`;

export const Border = styled.div`
  height: 1px;
  width: 100%;
  background: ${({ theme }) => theme.palette.gray.main};
`;
