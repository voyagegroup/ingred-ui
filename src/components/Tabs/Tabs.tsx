import * as React from "react";
import * as Styled from "./styled";
import { Tab } from "./internal/Tab";
import useEventCallback from "../../hooks/useEventCallback";
import { useMergeRefs } from "../../hooks/useMergeRefs";

type TabsProps<T> = {
  data: {
    text: string;
    count?: number;
    value: T;
  }[];
  value: T;
  withBadge?: boolean;
  onChange?: (value: T) => void;
  onClick?: (event: React.MouseEvent<HTMLButtonElement>) => void;
};

const Tabs = <T,>(
  props: TabsProps<T>,
  ref: React.Ref<HTMLDivElement>,
): React.ReactElement<TabsProps<T>> => {
  const { data, value, withBadge = false, onChange, onClick } = props;
  const valueToIndex = new Map<T, number>();
  const childrenRef = React.useRef<HTMLDivElement>(null);
  const [indicatorStyle, setIndicatorStyle] = React.useState({});
  const [tabs, setTabs] = React.useState<HTMLDivElement | null>(null);
  const refs = useMergeRefs<HTMLDivElement>(ref, setTabs);

  const getTabsMeta = () => {
    const tabsNode = tabs ?? undefined;
    let tabsMeta;
    if (tabsNode) {
      const rect = tabsNode.getBoundingClientRect();
      tabsMeta = {
        scrollLeft: tabsNode.scrollLeft,
        left: rect.left,
      };
    }

    let tabMeta;
    if (tabsNode && value) {
      if (childrenRef.current) {
        const children = childrenRef.current.children;

        if (children.length > 0) {
          const index = valueToIndex.get(value) as number;
          const tab = children[index];
          tabMeta = tab?.getBoundingClientRect() || null;
        }
      }
    }
    return { tabsMeta, tabMeta };
  };

  const updateIndicatorState = useEventCallback(() => {
    const { tabsMeta, tabMeta } = getTabsMeta();
    let startValue = 0;

    if (tabMeta && tabsMeta) {
      const correction = tabsMeta.scrollLeft;
      startValue = tabMeta.left - tabsMeta.left + correction;
    }

    const newIndicatorStyle = {
      left: startValue,
      width: tabMeta ? tabMeta.width : 0,
    };

    if (isNaN(indicatorStyle["left"]) || isNaN(indicatorStyle["width"])) {
      setIndicatorStyle(newIndicatorStyle);
    } else {
      const start = Math.abs(
        indicatorStyle["left"] - newIndicatorStyle["left"],
      );
      const size = Math.abs(
        indicatorStyle["width"] - newIndicatorStyle["width"],
      );

      if (start >= 1 || size >= 1) {
        setIndicatorStyle(newIndicatorStyle);
      }
    }
  });

  React.useEffect(() => {
    updateIndicatorState();
  });

  React.useEffect(() => {
    const handleResize = () => {
      updateIndicatorState();
    };

    window.addEventListener("resize", handleResize);
    return () => window.removeEventListener("resize", handleResize);
  }, [updateIndicatorState]);

  return (
    <Styled.ChildContainer ref={refs}>
      <Styled.Indicator
        style={{
          ...indicatorStyle,
        }}
      />
      <Styled.ChildrenContainer ref={childrenRef}>
        {data.map((d, index) => {
          const childValue = d.value;
          const selected = childValue === value;
          valueToIndex.set(childValue, index);
          return (
            <Tab
              key={index.toString()}
              selected={selected}
              value={childValue}
              count={d.count}
              withBadge={withBadge}
              text={d.text}
              onChange={onChange}
              onClick={onClick}
            />
          );
        })}
        <Styled.Border />
      </Styled.ChildrenContainer>
    </Styled.ChildContainer>
  );
};

export default React.forwardRef(Tabs) as <T>(
  props: TabsProps<T>,
  ref: React.ForwardedRef<HTMLDivElement>,
) => ReturnType<typeof Tabs>;
