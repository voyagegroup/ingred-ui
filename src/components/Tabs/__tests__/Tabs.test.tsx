import * as React from "react";
import "@testing-library/jest-dom/extend-expect";
import { cleanup } from "@testing-library/react";
import Tabs from "../";
import { renderWithThemeProvider } from "../../../utils/renderWithThemeProvider";

describe("Tabs component testing", () => {
  afterEach(cleanup);

  test("Tabs", () => {
    const options = {
      data: [
        { text: "hoge", value: "hoge" },
        { text: "fuga", value: "fuga" },
      ],
    };
    const { asFragment } = renderWithThemeProvider(
      <Tabs {...options} value={"hoge"} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("Tabs with badge", () => {
    const options = {
      data: [
        { text: "hoge", count: 1, value: "hoge" },
        { text: "fuga", count: 2, value: "fuga" },
        { text: "piyo", count: 3, value: "piyo" },
      ],
    };
    const { asFragment } = renderWithThemeProvider(
      <Tabs {...options} value={"hoge"} withBadge={true} />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
