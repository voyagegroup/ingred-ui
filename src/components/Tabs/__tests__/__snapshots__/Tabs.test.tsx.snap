// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tabs component testing Tabs 1`] = `
<DocumentFragment>
  <div
    class="sc-gsDKAQ dFzOKQ"
  >
    <div
      class="sc-jRQBWg bMRicj"
      style="left: 0px; width: 0px;"
    />
    <div
      class="sc-dkPtRN"
    >
      <button
        class="sc-hKwDye hWgBEh"
        tabindex="0"
        value="hoge"
      >
        <div
          class="sc-furwcr jaYMeI"
          display="flex"
        >
          <p
            class="sc-bdvvtL euPFjW sc-eCImPb dafOHH"
            color="#041C33"
            font-size="14px"
          >
            hoge 
          </p>
        </div>
      </button>
      <button
        class="sc-hKwDye ctPsiR"
        tabindex="-1"
        value="fuga"
      >
        <div
          class="sc-furwcr jaYMeI"
          display="flex"
        >
          <p
            class="sc-bdvvtL euPFjW sc-eCImPb euALer"
            color="#041C33"
            font-size="14px"
          >
            fuga 
          </p>
        </div>
      </button>
      <div
        class="sc-gKclnd coYCXA"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`Tabs component testing Tabs with badge 1`] = `
<DocumentFragment>
  <div
    class="sc-gsDKAQ dFzOKQ"
  >
    <div
      class="sc-jRQBWg bMRicj"
      style="left: 0px; width: 0px;"
    />
    <div
      class="sc-dkPtRN"
    >
      <button
        class="sc-hKwDye hWgBEh"
        tabindex="0"
        value="hoge"
      >
        <div
          class="sc-furwcr eLUTHW"
          display="flex"
        >
          <p
            class="sc-bdvvtL euPFjW sc-eCImPb dafOHH"
            color="#041C33"
            font-size="14px"
          >
            hoge 
          </p>
          <span
            class="sc-iCfMLu plEQf"
            color="#FFFFFF"
            font-size="0.65em"
            font-weight="bold"
            type="pill"
          >
            1
          </span>
        </div>
      </button>
      <button
        class="sc-hKwDye ctPsiR"
        tabindex="-1"
        value="fuga"
      >
        <div
          class="sc-furwcr eLUTHW"
          display="flex"
        >
          <p
            class="sc-bdvvtL euPFjW sc-eCImPb euALer"
            color="#041C33"
            font-size="14px"
          >
            fuga 
          </p>
          <span
            class="sc-iCfMLu jyvuZM"
            color="#FFFFFF"
            font-size="0.65em"
            font-weight="bold"
            type="pill"
          >
            2
          </span>
        </div>
      </button>
      <button
        class="sc-hKwDye ctPsiR"
        tabindex="-1"
        value="piyo"
      >
        <div
          class="sc-furwcr eLUTHW"
          display="flex"
        >
          <p
            class="sc-bdvvtL euPFjW sc-eCImPb euALer"
            color="#041C33"
            font-size="14px"
          >
            piyo 
          </p>
          <span
            class="sc-iCfMLu jyvuZM"
            color="#FFFFFF"
            font-size="0.65em"
            font-weight="bold"
            type="pill"
          >
            3
          </span>
        </div>
      </button>
      <div
        class="sc-gKclnd coYCXA"
      />
    </div>
  </div>
</DocumentFragment>
`;
